File: src/public/ts/Gaveteiro.ts
import HVC from "./HVC";
import PortaCartoes from "./PortaCartoes";

export default class Gaveteiro {
  
    private restritos: number[] = [];
    private ultimoRestrito:number = 0;
    private gavetas: string[];

    public carga(portaCartao:PortaCartoes){

      let len = portaCartao.conteudo.length
      let i = 0;
      let final = false;

      while(i < len && !final){
        
        let cartao = portaCartao.conteudo.shift()

        if (cartao){
          if(this.registrar(i, cartao) == "erro"){
            return "erro"
          }
          this.ultimoRestrito = i;
        }
        else{
          final = true;
        }
        if(cartao === "000"){
          final = true;
        }

        i++;
      } 
      
      return "sucesso"
    }
  
    public registrar(endereco: number, valor: string) {
      
      if(endereco + 1 >= HVC.config.MAX_GAVETA) {

        HVC.terminal.addError("\nErro de sobrecarga da pilha, limite de 100 registros")
        return "erro"

      } else {

        if(endereco < this.ultimoRestrito){
            const conteudo = this.ler(endereco);
            HVC.terminal.addError(`\nErro tentativa de sobrescrita de gaveta que armazena código fonte\nconteúdo da gaveta(${endereco}): ${conteudo}\n`);
            return "erro";
        }
    
        this.gavetas[endereco] = valor;

      }

      return "sucesso"
    }
  
    public ler(endereco: number){

      if (endereco < 0 || endereco > HVC.config.MAX_GAVETA || !this.gavetas[endereco]) {                                   
        HVC.terminal.addError(`\nErro na leitura do gaveteiro no endereço ${endereco}, tentativa de leitura em endereço inexistente\n`);
        return "erro";
      }
      
  
      return this.gavetas[endereco] as string;
    }

    public updateGaveteiro() {

      this.gavetas = []

    }

    public getRestByInstruct(inst:string) {

      return this.gavetas.indexOf(inst)

    }

  }
  

File: src/hvm/Gaveteiro.cpp
#include <stdexcept>
#include <iostream>
#include "hvm/Gaveteiro.hpp"
#include "hvm/PortaCartoes.hpp"

Gaveteiro::Gaveteiro(int maxGavetas) {
    this->gavetas = std::vector<std::string>(maxGavetas);
    this->ultimoRestrito = 0;
}

std::vector<std::string> Gaveteiro::getGavetas() {
    return this->gavetas;
}

void Gaveteiro::carga(PortaCartoes& portaCartao) {
    
    int index = 0;
    bool final = false;

    while (!final) {

        std::string cartao = portaCartao.lerCartao();

        if (!cartao.empty()) {
            this->registrar(index, cartao);
            final = cartao == "000";
        } else {
            throw std::runtime_error("Falha na carga do porta cartões para gaveteiro");
        }

        index++;
    }

    this->ultimoRestrito = index;
}

void Gaveteiro::registrar(int endereco, const std::string& valor) {

    if (endereco + 1 >= this->gavetas.size()) {
        throw std::runtime_error("Erro de sobrecarga de gavetas, limite de " + std::to_string(this->gavetas.size()) + " registros");
    } else {
        if (endereco < this->ultimoRestrito && this->ultimoRestrito > 0) {
            const std::string& conteudo = ler(endereco);
            throw std::runtime_error("Erro tentativa de sobrescrita de gaveta que armazena código fonte conteúdo da gaveta(" + std::to_string(endereco) + "): " + conteudo);
        }

        int numericValue = std::stoi(valor);

        if (numericValue < 0 || numericValue > 999) {
            throw std::runtime_error("Valor inválido de escrita na gaveta [" + std::to_string(endereco) + "]: " + std::to_string(numericValue));
        }
    }

    this->gavetas[endereco] = valor;
}

std::string Gaveteiro::ler(int endereco) {
    if (endereco < 0 || endereco >= this->gavetas.size() || this->gavetas[endereco].empty()) {
        throw std::runtime_error("Erro na leitura do gaveteiro no endereço " + std::to_string(endereco) + ", tentativa de leitura em endereço inexistente");
    }

    return this->gavetas[endereco];
}
